### PR TITLE
fixing memory leaks

### PR DESCRIPTION
**Pull request by Grinkers**
This is still very much WIP. I wanted to get feedback before proceeding (I'm only testing against `examples/shutdown`, no other runs, apps, etc). At least one issue is browsers send the `keep-alive` header and now using a browser will cause a pretty long timeout before exiting.

Using `cargo build --release --manifest-path examples/shutdown/Cargo.toml && valgrind --leak-check=full -- examples/shutdown/target/release/shutdown` on Fedora 39 (x86_64) + rustc 1.75 here is what I'm seeing. Testing with `curl localhost:8080/hello`, no other header info.

# Master
## 0 curls
```
==1121279== LEAK SUMMARY:
==1121279==    definitely lost: 0 bytes in 0 blocks
==1121279==    indirectly lost: 0 bytes in 0 blocks
==1121279==      possibly lost: 9,504 bytes in 33 blocks
==1121279==    still reachable: 10,698 bytes in 195 blocks
==1121279==         suppressed: 0 bytes in 0 blocks
```
## 5 curls
```
==1123754== LEAK SUMMARY:
==1123754==    definitely lost: 0 bytes in 0 blocks
==1123754==    indirectly lost: 0 bytes in 0 blocks
==1123754==      possibly lost: 9,504 bytes in 33 blocks
==1123754==    still reachable: 11,550 bytes in 217 blocks
==1123754==         suppressed: 0 bytes in 0 blocks
```

# grinkers/memory branch (this PR as of writing this)
## 0 curls
```
==1124996== HEAP SUMMARY:
==1124996==     in use at exit: 0 bytes in 0 blocks
==1124996==   total heap usage: 417 allocs, 417 frees, 33,002 bytes allocated
==1124996==
==1124996== All heap blocks were freed -- no leaks are possible
```
## 5 curls
```
==1124830== HEAP SUMMARY:
==1124830==     in use at exit: 0 bytes in 0 blocks
==1124830==   total heap usage: 683 allocs, 683 frees, 89,415 bytes allocated
==1124830==
==1124830== All heap blocks were freed -- no leaks are possible
```

At least a partial reproduction can be also seen with a small application like
```rust
fn main() -> Result<(), Box<dyn Error>> {
    let monitor = MonitorConfig::default();
    let mut thread_pool = ThreadPool::new(2);
    thread_pool.register_monitor(monitor);
    thread_pool.start();
    thread_pool.stop();
    drop(thread_pool);
    sleep(Duration::from_secs(5));
    Ok(())
}
```

### CODE REVIEW

**Comment by Grinkers** [comment]
I'm seeing similar results on Windows too. The same `cargo build --release --manifest-path examples/shutdown/Cargo.toml` using https://drmemory.org/

I think there's a bunch of other false positives of other issues in drmemory. I'll have to mess around with some other tooling before this gets fully resolved, but for now it's at least similar.

# master branch
```
      1 unique,     1 total,     32 byte(s) of leak(s)
      2 unique,     2 total,   2720 byte(s) of possible leak(s)
```

# grinkers/memory branch
```
      1 unique,     1 total,     32 byte(s) of leak(s)
      0 unique,     0 total,      0 byte(s) of possible leak(s)
 ```

**Comment by Grinkers** [comment]
For completion, here's 4 logs. master/grinkers on valgrind/drmemory (no curls on all of these).

[valgrind_master_results.txt](https://github.com/w-henderson/Humphrey/files/13844172/valgrind_master_results.txt)
[valgrind_grinkers_results.txt](https://github.com/w-henderson/Humphrey/files/13844173/valgrind_grinkers_results.txt)
[drmemory_grinkers_results.txt](https://github.com/w-henderson/Humphrey/files/13844174/drmemory_grinkers_results.txt)
[drmemory_master_results.txt](https://github.com/w-henderson/Humphrey/files/13844175/drmemory_master_results.txt)

**Comment by w-henderson** [comment]
Okay, can we go back a step - do you have any ideas as to what in the code is causing these potential memory leaks?

**Comment by Grinkers** [comment]
> Okay, can we go back a step - do you have any ideas as to what in the code is causing these potential memory leaks?

TLDR: `ThreadPool` threads need a way to exit their loops on all threads, so drop can be called on all the things (RAII).

Sure. To be clear, "memory leak" might be too strong. You can see in the logs and such that it is classified as "possibly lost".  What I think is happening is the threadpool isn't breaking out of its `loop`s. They end up never dropping their internals, so if you stop and start a bunch of servers, you'll end up with leaks. In the OP, you can see a small rust program that does a replication of the main issue, creating, starting, stopping, and then dropping the `ThreadPool`. Please keep in mind I've only researched this as far as TheadPool and the code that is exercised in `examples/shutdown`. I don't know the implications beyond this.

In pseudo code this would cause this to happen, but is still fully defined (this sort of leaking is considered safe rust. there's never a use before initialization, use after free, or double free). in this example, using c's malloc. but anything that doesn't get its destructor/drop called.
```rust
thread::spawn(|| {
  m = malloc();
  loop {}
  free(m);
)
```
The pointer m is still available, so it is "possibly lost" -- IE we still have the pointer `m` to call free on, but we never do. There's also some allocations that rust does when doing `spawn`, which is why I added https://github.com/w-henderson/Humphrey/pull/114/files?diff=unified&w=1#diff-0c830b90b3ef1ce63ac62ac061bf83dcbad03d919c5504617d91df4621e49135R25. Waiting a couple seconds would get rid of this positive hit. I believe there's no issues with full memory leaks, as rust is generally really good about calling free on all drops, unless explicitly leaked.
```rust
m = malloc();
let _ = mem::forget(m); // this would be a real leak
```

**Comment by Grinkers** [comment]
Only supplying on linux, as it's easier. Can research more on Windows later (but I suspect it's the same).

minimal reproduction/example. Attached are two runs, with a different number of loops (1 and 5). On `Grinkers:memory`, there's no leaks possible. 
[valgrind_minimal_repo_1_loop.txt](https://github.com/w-henderson/Humphrey/files/13848048/valgrind_minimal_repo_1_loop.txt)
[valgrind_minimal_repo_5_loop.txt](https://github.com/w-henderson/Humphrey/files/13848049/valgrind_minimal_repo_5_loop.txt)

```rust
use humphrey::monitor::MonitorConfig;
use humphrey::thread::pool::ThreadPool;

use std::error::Error;
use std::thread::sleep;
use std::time::Duration;

fn i_leak() {
    let monitor = MonitorConfig::default();
    let mut thread_pool = ThreadPool::new(2);
    thread_pool.register_monitor(monitor);
    thread_pool.start();
    thread_pool.stop();
    drop(thread_pool);
}

fn main() -> Result<(), Box<dyn Error>> {
    for _ in 0..1 {
        // changing the number of runs changes the amount still in use
        i_leak();
    }

    sleep(Duration::from_secs(5)); // plenty of time for threads that would exit, to actually exit
    Ok(())
}
```

File: humphrey/src/thread/pool.rs [hunk 1]
```diff
@@ -104,8 +105,12 @@ impl ThreadPool {
 
     /// Stops the thread pool.
     pub fn stop(&mut self) {
-        self.recovery_thread = None;
-        self.tx.send(Message::Shutdown).unwrap();
+        self.tx = None;
```
Comment (w-henderson) [inline humphrey/src/thread/pool.rs:108]: I'm probably missing something obvious, but if we're no longer sending `Message::Shutdown`, what breaks the threads out of their loops?

Comment (Grinkers) [reply humphrey/src/thread/pool.rs:111]: https://github.com/w-henderson/Humphrey/pull/114/files#diff-bbf89d851fd088ede1c3526bb6e2bdf95406f4cb9de43656ef9602766c6dd08fR66
When the tx gets dropped, the `break 'outer;` fires. 

btw I'm very much not convinced with this PR's implementation and wanted feedback. I only briefly researched this in the context of the shutdown example and don't know the implications for all the crate features. In the near future, I plan to start using websockets, so will be messing around with that and will check memory stuff at the same time. I think there might be some weirdness there because of connections staying open?

On a side note, was it intentional to have ThreadPool pub?

Comment (w-henderson) [reply humphrey/src/thread/pool.rs:111]: I was referring to the regular (not recovery threads). I don't think `ThreadPool` was really supposed to be public - honestly this whole part of the project is very dodgy.

Comment (Grinkers) [reply humphrey/src/thread/pool.rs:111]: It's not fresh on my mind, but I don't think there was a problem with regular threads. The drop impl handled it. In this WIP I changed it to be join so it would wait for them to exit for the memory checkers, but this might not be great behavior. As I mentioned in the OP, the `keep-alive` header makes this join take a long time
https://github.com/w-henderson/Humphrey/pull/114/files#diff-70f90e648844ce54d1a7dde834fbaa03af4128e594489fd77631137e852f88e8R208

Comment (Grinkers) [reply humphrey/src/thread/pool.rs:111]: https://github.com/Grinkers/Humphrey/blob/8324254edc2f3f4ca6973d93d71a4762b3fa3f4a/humphrey/src/thread/pool.rs#L162
I later realized I didn't really answer the question, regular threads also break with a similar pattern. I made a small push to have an explicit send too, although not strictly needed.

Comment (w-henderson) [reply humphrey/src/thread/pool.rs:111]: Okay, that makes sense - can you explain why `self.tx` needs to be an `Option`?

Comment (Grinkers) [reply humphrey/src/thread/pool.rs:111]: I needed to be able to drop `tx` to be able to break out of this iterator (which would otherwise wait forever)
https://github.com/w-henderson/Humphrey/pull/114/files#diff-bbf89d851fd088ede1c3526bb6e2bdf95406f4cb9de43656ef9602766c6dd08fR65

link to docs for behavior
https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.iter

One of the reasons I asked if it was intentional to be pub is because these things are actually breaking changes. If it was just internal, it wouldn't matter.

**Comment by Grinkers** [comment]
I started messing with async websockets, and this PR gives no possible leaks.

I think the biggest thing right now is figuring out what to do with the 'keep-alive` header. With the current implementation
https://github.com/w-henderson/Humphrey/commit/d55f5e92f8d15a43a2399d3fd7c0231c89535d5b#diff-70f90e648844ce54d1a7dde834fbaa03af4128e594489fd77631137e852f88e8L201-R208
ends up blocking until this
https://github.com/w-henderson/Humphrey/blob/d07e3ff71cfbc49094edbaa518480ec7ebeef693/humphrey/src/app.rs#L569-L572
times out or exits.
 
I can make it `drop(thread)` again. The App will return almost immediately, but then it'll show memory leaks if the executable exits before the timeout/return. Also there might be some weird behavior with trying to `TcpListener::bind` while a `keep-alive` is still looping?

I suggest we keep `join()`, as it's more correct. Normal use cases (long-living applications) can manually put the App in a thread and check the JoinHandle themselves, and log/wait/whatever. With `drop`, all information is lost. 

From a mocking/testing perspective (multiple runs/stops and memory checkers), it's easy enough to just add `.with_connection_timeout(Some(Duration::from_secs(1)))`, or some relatively small number.

Another alternative is to make TcpStream [nonblocking](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_nonblocking) (which ends up needing platform specific handling) or use channels (performance hit via locks), but that seems like a lot of work for such an edge case. Please tell me if you have any other ideas.

**Comment by Grinkers** [comment]
bump. 

I've been using this branch, while regularly looking for memory leaks, using `App` and `AsyncWebsocketApp` without issues or leaks. In response to my last message, I think the behavior in this branch as-is, is likely the best given the expected workflows.

**Comment by w-henderson** [comment]
Sorry for leaving this for so long and thank you again so much for your work on this - I have exams in a month but after that I'll give it a proper review and get it merged.

**Comment by Grinkers** [comment]
No worries at all! I'm totally fine using my branch.

Good luck with your exams.
